### PR TITLE
[WIP] Initial code for OpenEBS CSI Plugin

### PR DESCRIPTION
**Pull request by princerachit**
This is a WIP PR. Please <B>DO NOT MERGE</B>.
[Design doc](https://docs.google.com/document/d/16xTjQgLc2AqGMhunbv3ZPw-EkjjW4F3UPVMzJ8HkqXQ/edit?usp=sharing)
<del>Some files are not tested and </del>some cannot be tested.

Ignore the following files as they have been copied from `kubernetes-csi/drivers` repo and thus do not need to be tested. These files maybe removed in future and instead a common library could be used to perform its job.

1. iscsi.go
2. iscsi_util.go
3. server.go
4. driver.go

<del>Leveled logs would be added soon.</del>

### CODE REVIEW

**Comment by princerachit** [comment]
image

File: pkg/openebs/mayaproxy/mayaproxy.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	responseBody := string(data)
+	glog.Infof("volume Successfully Created:\n%v\n", responseBody)
+	return nil
+}
+
+// DeleteVolume requests mapi server to delete an openebs volume. It returns an error if volume deletion fails
+func (mayaService *MayaService) DeleteVolume(mapiURI *url.URL, volumeName string) error {
+	glog.Infof("[DEBUG] Delete Volume :%v", string(volumeName))
```
Comment (prateekpandey14) [inline pkg/openebs/mayaproxy/mayaproxy.go]: better to use log level here for debug logs, for example `glog.V(2).Infof("----")` .

File: pkg/openebs/mayaproxy/mayaproxy.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+// of a volume and returns it by filling into *mayav1.Volume.
+// If the volume does not exist or can't be retrieved then it returns an error
+func (mayaService *MayaService) GetVolume(mapiURI *url.URL, volumeName string) (*mayav1.Volume, error) {
+	var volume mayav1.Volume
+
+	url, err := mayaService.GetVolumeInfoURL(mapiURI, versionLatest, volumeName)
+	if err != nil {
+		return nil, err
+	}
+	glog.Infof("[DEBUG] Requesting for volume details at %s", url.String())
```
Comment (prateekpandey14) [inline pkg/openebs/mayaproxy/mayaproxy.go]: Same here.

File: pkg/openebs/mayaproxy/mayaproxy.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	glog.Infof("volume Details Successfully Retrieved %v", volume)
+
+	return &volume, nil
+}
+
+// ListAllVolumes requests mapi server to GET the details of all volumes
+func (mayaService *MayaService) ListAllVolumes(mapiURI *url.URL) (*[]mayav1.Volume, error) {
+	var volumesList mayav1.VolumeList
+
+	glog.Infof("[DEBUG] Get details for all volumesList")
```
Comment (prateekpandey14) [inline pkg/openebs/mayaproxy/mayaproxy.go]: same as above.

File: pkg/openebs/openebs.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (prateekpandey14) [inline pkg/openebs/openebs.go:2]: nit : OpenEBS ?

File: pkg/openebs/server/controllerserver.go [hunk 5]
```diff
@@ -0,0 +1,252 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (prateekpandey14) [inline pkg/openebs/server/controllerserver.go:2]: nit: OpenEBS ?

File: pkg/openebs/server/controllerserver.go [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+	"github.com/princerachit/csi-openebs/pkg/openebs/driver"
+	"strconv"
+	"strings"
+	"errors"
+)
+
+var (
+	mayaConfig        *mayaproxy.MayaConfig
+	clientWrapper     *mayaproxy.K8sClientWrapper
+	mayaConfigBuilder mayaproxy.Builder
```
Comment (prateekpandey14) [inline pkg/openebs/server/controllerserver.go:39]: Please explain these var , some precise description ?

File: pkg/openebs/server/controllerserver.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+)
+
+type ControllerServer struct {
```
Comment (prateekpandey14) [inline pkg/openebs/server/controllerserver.go]: Explain this struct please.

File: pkg/openebs/server/controllerserver.go [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+	csi.ControllerServer
+	Driver *driver.CSIDriver
+}
+
+func checkArguments(req *csi.CreateVolumeRequest) error {
```
Comment (prateekpandey14) [inline pkg/openebs/server/controllerserver.go]: Add some function level comments too.

File: pkg/openebs/server/controllerserver.go [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+			return nil, status.Error(codes.Unavailable, fmt.Sprint(err))
+		}
+	}
+
+	volume, err = mayaConfig.MayaService.GetVolume(&mayaConfig.MapiURI, req.GetName())
+	if err != nil {
+		return nil, status.Error(codes.DeadlineExceeded, fmt.Sprintf("Unable to contact mapi server: %v", err))
+	}
+
+	glog.Infof("[DEBUG] Volume details %s", volume)
```
Comment (prateekpandey14) [inline pkg/openebs/server/controllerserver.go]: please use log levels , explained in above comment.

File: pkg/openebs/server/controllerserver.go [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+	if mayaConfig == nil {
+		mayaConfig, err = mayaConfigBuilder.GetNewMayaConfig(clientWrapper)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
```
Comment (prateekpandey14) [inline pkg/openebs/server/controllerserver.go]: Add some method level comments here.

File: pkg/openebs/server/identityserver.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (prateekpandey14) [inline pkg/openebs/server/identityserver.go:2]: nit : OpenEBS ?

File: pkg/openebs/server/iscsi.go [hunk 12]
```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (prateekpandey14) [inline pkg/openebs/server/iscsi.go:2]: nit : OpenEBS ?

File: pkg/openebs/server/iscsi.go [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/container-storage-interface/spec/lib/go/csi/v0"
+	"k8s.io/kubernetes/pkg/util/mount"
+	"k8s.io/kubernetes/pkg/volume/util"
+)
+
+func getISCSIInfo(req *csi.NodePublishVolumeRequest) (*iscsiDisk, error) {
```
Comment (prateekpandey14) [inline pkg/openebs/server/iscsi.go]: Add some function level comments here.

File: pkg/openebs/server/nodeserver.go [hunk 14]
```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (prateekpandey14) [inline pkg/openebs/server/nodeserver.go:2]: nit : OpenEBS ?

File: pkg/openebs/server/nodeserver.go [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+	"github.com/princerachit/csi-openebs/pkg/openebs/driver"
+)
+
+type NodeServer struct {
+	csi.NodeServer
+	Driver *driver.CSIDriver
+}
+
+func (ns *NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolumeRequest) (*csi.NodePublishVolumeResponse, error) {
```
Comment (prateekpandey14) [inline pkg/openebs/server/nodeserver.go]: Add some function level comments here, as well as code level comments if requires.

File: pkg/openebs/server/server.go [hunk 16]
```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (prateekpandey14) [inline pkg/openebs/server/server.go:2]: nit : OpenEBS ?

File: pkg/openebs/server/server.go [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+*/
+
+package server
+
+import (
+	"net"
+	"os"
+	"sync"
+
+	"github.com/container-storage-interface/spec/lib/go/csi/v0"
```
Comment (prateekpandey14) [inline pkg/openebs/server/server.go]: v0 ? latest is v2, check please. Do we have any limitations here with latest version ?

File: pkg/openebs/server/server.go [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+	return &nonBlockingGRPCServer{}
+}
+
+// NonBlocking server
+type nonBlockingGRPCServer struct {
+	wg     sync.WaitGroup
+	server *grpc.Server
+}
+
+func (s *nonBlockingGRPCServer) Start(endpoint string, ids csi.IdentityServer, cs csi.ControllerServer, ns csi.NodeServer) {
```
Comment (prateekpandey14) [inline pkg/openebs/server/server.go:54]: Add method level comments here.

File: pkg/openebs/server/server.go [hunk 19]
```diff
@@ -0,0 +1,136 @@
...
+
+func (s *nonBlockingGRPCServer) Stop() {
+	s.server.GracefulStop()
+}
+
+func (s *nonBlockingGRPCServer) ForceStop() {
+	s.server.Stop()
+}
+
+func (s *nonBlockingGRPCServer) serve(endpoint string, ids csi.IdentityServer, cs csi.ControllerServer, ns csi.NodeServer) {
```
Comment (prateekpandey14) [inline pkg/openebs/server/server.go:75]: Here too.

File: pkg/openebs/v1/types.go [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+	Status struct {
+		Message string `json:"Message"`
+		Phase   string `json:"Phase"`
+		Reason  string `json:"Reason"`
+	} `json:"Status"`
+	Metadata struct {
+		Annotations       interface{} `json:"annotations"`
+		CreationTimestamp interface{} `json:"creationTimestamp"`
+		Name              string      `json:"name"`
+	} `json:"metadata"`
```
Comment (prateekpandey14) [inline pkg/openebs/v1/types.go:61]: We have to clean this volume struct schema and arrange them properly.

File: Gopkg.toml [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+[[constraint]]
+  name = "google.golang.org/grpc"
+  version = "1.10.0"
+
+[[constraint]]
+  name = "gopkg.in/gcfg.v1"
+  version = "1.2.1"
+
+[[constraint]]
+  version = "kubernetes-1.10.0-beta.1"
```
Comment (prateekpandey14) [inline Gopkg.toml:65]: Use stable version instead of beta.

File: pkg/openebs/server/nodeserver.go [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANYcon KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"github.com/container-storage-interface/spec/lib/go/csi/v0"
```
Comment (prateekpandey14) [inline pkg/openebs/server/nodeserver.go:20]: v0 ? does v0 get the latest version ?

File: pkg/openebs/server/identityserver.go [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"github.com/openebs/csi-openebs/pkg/openebs/driver"
+	"github.com/container-storage-interface/spec/lib/go/csi/v0"
```
Comment (prateekpandey14) [inline pkg/openebs/server/identityserver.go]: v0 ? latest is v2, check please. Do we have any limitations here with latest version ?

File: pkg/openebs/driver/driver_test.go [hunk 24]
```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (prateekpandey14) [inline pkg/openebs/driver/driver_test.go:2]: nit: OpenEBS ?

File: pkg/openebs/openebs.go (hunk 4 above)
Comment (princerachit) [reply pkg/openebs/openebs.go:2]: What should we put here? Only replace the Kubernetes or something different should go her altogether?

File: pkg/openebs/server/nodeserver.go (hunk 22 above)
Comment (princerachit) [reply pkg/openebs/server/nodeserver.go:20]: will look into it.

**Comment by princerachit** [comment]
@prateekpandey14 I will address the logging issues at last. If you could look into test cases and their messages then it would be helpful.

File: pkg/openebs/server/server.go (hunk 19 above)
Comment (princerachit) [reply pkg/openebs/server/server.go:75]: This is a borrowed code from kubernetes-csi. I will add comments anyway.

File: Makefile [hunk 25]
```diff
@@ -0,0 +1,25 @@
+# Copyright 2017 The Kubernetes Authors.
```
Comment (prateekpandey14) [inline Makefile:1]: nit : OpenEBS ?

File: example/csi-setup.yaml [hunk 26]
```diff
@@ -0,0 +1,247 @@
...
+kind: ClusterRole
+metadata:
+  name: csi-cluster-role
+rules:
+- apiGroups:
+  - ""
+  resources:
+  - persistentvolumes
+  verbs:
+  - create
```
Comment (prateekpandey14) [inline example/csi-setup.yaml:28]: arrange all the verbs in comma separated array.

File: example/csi-setup.yaml [hunk 27]
```diff
@@ -0,0 +1,247 @@
...
+            name: socket-dir
+        - name: openebs-driver
+          args:
+          - --endpoint=$(CSI_ENDPOINTS)
+          env:
+          - name: "OPENEBS_NAMESPACE"
+            value: "default"
+          - name: CSI_ENDPOINTS
+            value: unix:///var/lib/csi/sockets/pluginproxy/csi.sock
+          image: rachitprince/csi-openebs:v1.0.0
```
Comment (prateekpandey14) [inline example/csi-setup.yaml:236]: nit: change to openebs/csi-openebs ?

File: pkg/openebs/driver/driver.go [hunk 28]
```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (prateekpandey14) [inline pkg/openebs/driver/driver.go:2]: Add OpenEBS  too ?

File: Gopkg.toml (hunk 21 above)
Comment (princerachit) [reply Gopkg.toml:65]: Things are breaking when I move away from beta. Will resolve this later.

**Review by kmova — CHANGES_REQUESTED** [review]
Please provide the link to the design doc in the description.

**Comment by princerachit** [comment]
Done @kmova

**Comment by blaggacao** [comment]
@princerachit Why has this been closed? Is a replacement in the works?

**Comment by princerachit** [comment]
Hi,
I worked on this in 2018 when CSI spec were not evolved and lacked a lot of
information. This PR was put on hold as there were some big changes
parallel being made in openebs/maya and external provisioner.

I think a different team has recently already taken up this task. If you
have any other questions let me know.

**Comment by kmova** [comment]
@blaggacao .. the implementation is being moved to https://github.com/openebs/maya/tree/csi to keep all of the OpenEBS control plane work in one repository. The latest work is built on top of what @princerachit already did as a prototype in this repo!

**Pull request closed without merging**